// liter/lib/src/conversion-network.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var unit_1 = require("./unit");
var UnitConversionNetwork = /** @class */ (function () {
    function UnitConversionNetwork() {
        this.edges = {};
    }
    UnitConversionNetwork.prototype.addEdge = function (left, leftRate, right, rightRate) {
        if (!this.edges[left])
            this.edges[left] = {};
        if (this.edges[left][right])
            throw new Error('duplicate edge');
        this.edges[left][right] = { leftRate: leftRate, rightRate: rightRate };
    };
    UnitConversionNetwork.prototype.addRate = function (num, numUnit, denom, denomUnit) {
        this.rates = undefined;
        var numRate = numUnit.mul(unit_1.Unit.one(num));
        var denomRate = denomUnit.mul(unit_1.Unit.one(denom));
        this.addEdge(numUnit.termsKey(), numRate, denomUnit.termsKey(), denomRate);
        this.addEdge(denomUnit.termsKey(), denomRate, numUnit.termsKey(), numRate);
    };
    UnitConversionNetwork.prototype.computeRates = function () {
        var _this = this;
        if (this.rates)
            return this.rates;
        this.rates = {};
        var rates = this.rates;
        var walk = function (fromK, leftK, pathRate, visited) {
            for (var rightK in _this.edges[leftK]) {
                if (rightK === fromK || visited[rightK])
                    continue;
                visited[rightK] = true;
                var _a = _this.edges[leftK][rightK], leftRate = _a.leftRate, rightRate = _a.rightRate;
                var rate = pathRate.mul(leftRate.div(rightRate));
                if (rates[rate.termsKey()] != null)
                    throw new Error('duplicate rate');
                rates[rate.termsKey()] = rate;
                walk(fromK, rightK, rate, visited);
            }
        };
        for (var fromK in this.edges)
            walk(fromK, fromK, unit_1.Unit.one(), {});
        return rates;
    };
    UnitConversionNetwork.prototype.getConversionUnit = function (from, to) {
        var rates = this.computeRates();
        var rateU = to.div(from);
        return rates[rateU.termsKey()];
    };
    return UnitConversionNetwork;
}());
exports.UnitConversionNetwork = UnitConversionNetwork;
;
