// liter/lib/tests/units.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../index");
var chai_1 = require("chai");
require("mocha");
var epsilon = 1e-9;
var baseUnitNames = [
    'meter',
    'kilogram',
    'second',
    'ampere',
    'kelvin',
    'mole',
    'candela',
    'radian',
    'steradian',
];
describe('Unit basics', function () {
    it('should be creatable', function () {
        var props = {
            mode: 'normal',
            name: { test: 1 },
            terms: [1, 2, 3, 4, 5, 6, 7, 8, 9],
            customTerms: { test: 1 },
            m: [1, 0],
        };
        var u = index_1.Unit.one().set(props);
        chai_1.expect(u.mode).to.equal(props.mode);
        chai_1.expect(u.name).to.deep.equal(props.name);
        chai_1.expect(u.terms).to.deep.equal(props.terms);
        chai_1.expect(u.customTerms).to.deep.equal(props.customTerms);
        chai_1.expect(u.m).to.deep.equal(props.m);
    });
    it('should have base units pre-populated', function () {
        var term = 0;
        var terms = [0, 0, 0, 0, 0, 0, 0, 0, 0];
        for (var _i = 0, baseUnitNames_1 = baseUnitNames; _i < baseUnitNames_1.length; _i++) {
            var base = baseUnitNames_1[_i];
            var u = index_1.baseUnits[base];
            chai_1.expect(u).to.exist;
            chai_1.expect(u.nameFormat()).to.equal(base);
            chai_1.expect(u.baseFormat()).to.equal(base);
            terms[term] = 1;
            chai_1.expect(u.terms).to.deep.equal(terms);
            terms[term++] = 0;
            chai_1.expect(u.m).to.deep.equal([1, 0]);
        }
    });
    describe('Unit.one()', function () {
        it('should create a unit with appropriate m', function () {
            var u = index_1.Unit.one(42);
            chai_1.expect(u.m).to.deep.equal([42, 0]);
            u = index_1.Unit.one(5 / 9, 32);
            chai_1.expect(u.m).to.deep.equal([5 / 9, 32]);
        });
    });
    it('should add terms on multiplication', function () {
        var u = index_1.baseUnits.meter.mul(index_1.baseUnits.meter, index_1.baseUnits.kilogram, index_1.baseUnits.second);
        chai_1.expect(u).to.exist;
        chai_1.expect(u.nameFormat()).to.equal('meter^2 kilogram second');
        chai_1.expect(u.baseFormat()).to.equal('meter^2 kilogram second');
    });
    it('should subtract terms on division', function () {
        var u = index_1.baseUnits.meter.div(index_1.baseUnits.kilogram, index_1.baseUnits.second, index_1.baseUnits.second);
        chai_1.expect(u).to.exist;
        chai_1.expect(u.nameFormat()).to.equal('meter kilogram^-1 second^-2');
        chai_1.expect(u.baseFormat()).to.equal('meter kilogram^-1 second^-2');
    });
    it('should allow adding same unit', function () {
        chai_1.expect(function () { return index_1.baseUnits.meter.add(index_1.baseUnits.meter); }).to.not.throw(Error);
    });
    it('should not allow adding different unit', function () {
        chai_1.expect(function () { return index_1.baseUnits.meter.add(index_1.baseUnits.kilogram); }).to.throw('Can not add <kilogram> to <meter>');
    });
    it('should allow adding same unit', function () {
        chai_1.expect(function () { return index_1.baseUnits.meter.add(index_1.baseUnits.meter); }).to.not.throw();
        var _a = index_1.baseUnits.meter.add(index_1.baseUnits.meter), u = _a[0], om = _a[1], lm = _a[2], rm = _a[3];
        chai_1.expect(u).to.deep.equal(index_1.baseUnits.meter);
        for (var _i = 0, _b = [om, lm, rm]; _i < _b.length; _i++) {
            var m = _b[_i];
            chai_1.expect(m[0]).to.be.closeTo(1, epsilon);
            chai_1.expect(m[1]).to.be.closeTo(0, epsilon);
        }
    });
    describe('Custom bases', function () {
        it('should have correct name', function () {
            var u = index_1.Unit.base('frog');
            chai_1.expect(u.nameFormat()).to.equal('frog');
            chai_1.expect(u.baseFormat()).to.equal('frog');
        });
        it('should return false from isDimensionless()', function () {
            chai_1.expect(index_1.Unit.base('frog').isDimensionless()).to.be.false;
        });
    });
});
